src/main/java/com/google/devtools/build/lib/analysis/RepoMappingManifestAction.java

```diff
@@ -0,0 +1,122 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.google.devtools.build.lib.analysis;
+
+import static java.nio.charset.StandardCharsets.ISO_8859_1;
+
+import com.google.common.collect.ImmutableSortedMap;
+import com.google.common.collect.Maps;
+import com.google.devtools.build.lib.actions.ActionExecutionContext;
+import com.google.devtools.build.lib.actions.ActionKeyContext;
+import com.google.devtools.build.lib.actions.ActionOwner;
+import com.google.devtools.build.lib.actions.Artifact;
+import com.google.devtools.build.lib.actions.Artifact.ArtifactExpander;
+import com.google.devtools.build.lib.actions.CommandLineExpansionException;
+import com.google.devtools.build.lib.actions.ExecException;
+import com.google.devtools.build.lib.analysis.actions.AbstractFileWriteAction;
+import com.google.devtools.build.lib.analysis.actions.DeterministicWriter;
+import com.google.devtools.build.lib.cmdline.RepositoryMapping;
+import com.google.devtools.build.lib.cmdline.RepositoryName;
+import com.google.devtools.build.lib.collect.nestedset.NestedSetBuilder;
+import com.google.devtools.build.lib.collect.nestedset.Order;
+import com.google.devtools.build.lib.util.Fingerprint;
+import com.google.devtools.build.lib.vfs.OsPathPolicy;
+import java.io.BufferedWriter;
+import java.io.OutputStreamWriter;
+import java.io.Writer;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.UUID;
+import javax.annotation.Nullable;
+import net.starlark.java.eval.EvalException;
+
+/** Creates a manifest file describing the repos and mappings relevant for a runfile tree. */
+public class RepoMappingManifestAction extends AbstractFileWriteAction {
+  private static final UUID MY_UUID = UUID.fromString("458e351c-4d30-433d-b927-da6cddd4737f");
+
+  private final ImmutableSortedMap<RepositoryName, ImmutableSortedMap<String, RepositoryName>>
+      reposAndMappings;
+  private final String workspaceName;
+
+  public RepoMappingManifestAction(
+      ActionOwner owner,
+      Artifact output,
+      Map<RepositoryName, RepositoryMapping> reposAndMappings,
+      String workspaceName) {
+    super(owner, NestedSetBuilder.emptySet(Order.STABLE_ORDER), output, /*makeExecutable=*/ false);
+    this.reposAndMappings =
+        ImmutableSortedMap.copyOf(
+            Maps.transformValues(
+                reposAndMappings, repoMapping -> ImmutableSortedMap.copyOf(repoMapping.entries())),
+            (a, b) -> OsPathPolicy.getFilePathOs().compare(a.getName(), b.getName()));
```

Comment:
This is just to be consistent with the implementation of RepositoryName#equals, which also uses the OS-dependent thing. In practice this shouldn't really matter. (Is getting consistent outputs across OSes important?)

Reply:
It's not very important - cross-platform caching doesn't really work yet anyway since the output paths encode the CPU. But since I'm also working on improving that, I'm trying to minimize OS dependencies whenever I see them.

Reply:
I decided to scrap this, since it's awkward whether or not we sort OS-dependently, and it's better to be less verbose. I have half a mind to just disallow uppercase letters in repo names...